### PR TITLE
fix: fail early if index is too dimensional

### LINKED ISSUE #2302: Slicing `dask-awkward` array in too many dimensions raises `StopIteration`
### Description of new feature

Trying to slice an Awkward Array in too many dimensions raises a helpful error:

```python
>>> import awkward as ak, dask_awkward as dak
>>> a = ak.Array([1])
>>> a[:, 0]
Traceback (most recent call last):
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/numpyarray.py", line 346, in _getitem_next
    out = self._data[where]
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/highlevel.py", line 951, in __getitem__
    out = self._layout[where]
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/content.py", line 548, in __getitem__
    return self._getitem(where)
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/content.py", line 589, in _getitem
    out = next._getitem_next(nextwhere[0], nextwhere[1:], None)
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/regulararray.py", line 521, in _getitem_next
    nextcontent._getitem_next(nexthead, nexttail, advanced),
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/numpyarray.py", line 348, in _getitem_next
    raise ak._errors.index_error(self, (head, *tail), str(err)) from err
IndexError: while attempting to slice

    <Array [1] type='1 * int64'>

with

    (:, 0)

at inner NumpyArray of length 1, using sub-slice (0).

Error details: too many indices for array: array is 1-dimensional, but 2 were indexed.
```

Whereas for `dask-awkward` arrays, you get this much less friendly uncaught `StopIteration`, coming from `typetracer`:

```python
>>> da = dak.from_awkward(a, 1)
>>> da[:, 0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/iris-hep/src/dask-awkward/src/dask_awkward/lib/core.py", line 920, in __getitem__
    return self._getitem_tuple(where)
  File "/home/user/iris-hep/src/dask-awkward/src/dask_awkward/lib/core.py", line 847, in _getitem_tuple
    return self._getitem_trivial_map_partitions(where)
  File "/home/user/iris-hep/src/dask-awkward/src/dask_awkward/lib/core.py", line 722, in _getitem_trivial_map_partitions
    meta = self._meta[metad]
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/highlevel.py", line 951, in __getitem__
    out = self._layout[where]
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/content.py", line 548, in __getitem__
    return self._getitem(where)
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/content.py", line 589, in _getitem
    out = next._getitem_next(nextwhere[0], nextwhere[1:], None)
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/regulararray.py", line 521, in _getitem_next
    nextcontent._getitem_next(nexthead, nexttail, advanced),
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/contents/numpyarray.py", line 346, in _getitem_next
    out = self._data[where]
  File "/home/user/iris-hep/src/awkward-1.0/src/awkward/_nplikes/typetracer.py", line 456, in __getitem__
    dimension_length = next(iter_shape)
StopIteration
```

Ideally, this case should produce a message similar to the first example.

### PR DESCRIPTION
**Pull request by agoose77**
Fixes #2302 , by directly checking the minimum dimension of the array that is expected by the index.

### CODE REVIEW

**Review by jpivarski — APPROVED** [review]
I should mention that the `TypeTracerArray.__getitem__` only implements as many cases as we have in fact encountered; I remember that it wasn't fully general. Catching ellipsis errors early is another good step toward better coverage.

I'm not saying that it should be made fully general now (which would be hard), since we don't expose it to arbitrary slicing by users. I'm just saying that as testing of typetracers broadens, there might be additional cases to cover.

Looks good! As usual, when I use the "approve" button here, I'm meaning that it's ready to merge, but I don't merge it myself unless I'm 99% sure you're not making additional changes.

**Comment by agoose77** [comment]
> I should mention that the TypeTracerArray.__getitem__ only implements as many cases as we have in fact encountered; I remember that it wasn't fully general. Catching ellipsis errors early is another good step toward better coverage.
I'm not saying that it should be made fully general now (which would be hard), since we don't expose it to arbitrary slicing by users. I'm just saying that as testing of typetracers broadens, there might be additional cases to cover.

Actually, I think we have implemented all of the indexing types that a NumPy array supports. It's fairly long, although I think quite readable!

> Looks good! As usual, when I use the "approve" button here, I'm meaning that it's ready to merge, but I don't merge it myself unless I'm 99% sure you're not making additional changes.

Fab! In future, I'll explicitly tell you I'm hand's off, so that you get the fun of merging.

**Pull request merged**
